src/main.rs

```diff
@@ -0,0 +1,859 @@
+use crate::rdf::{
+    add_derived_from, add_five_star_annotation, add_property, add_quality_measurement,
+    convert_term_to_named_or_blank_node_ref, create_metrics_store, dump_graph_as_turtle,
+    get_dataset_node, has_property, is_rdf_format, list_distributions, list_formats,
+    list_media_types, parse_turtle,
+};
+use crate::schemas::{setup_schemas, DatasetEvent, DatasetEventType, MQAEvent, MQAEventType};
+
+use crate::reference_data::{valid_file_type, valid_media_type};
+
+mod rdf;
+mod reference_data;
+mod schemas;
+mod utils;
+mod vocab;
+
+async fn record_borrowed_message_receipt(msg: &BorrowedMessage<'_>) {
+    // Simulate some work that must be done in the same order as messages are
+    // received; i.e., before truly parallel processing can begin.
+    info!("Message received: {}", msg.offset());
+}
+
+async fn record_owned_message_receipt(_msg: &OwnedMessage) {
+    // Like `record_borrowed_message_receipt`, but takes an `OwnedMessage`
+    // instead, as in a real-world use case  an `OwnedMessage` might be more
+    // convenient than a `BorrowedMessage`.
+}
+
+fn parse_dataset_event(
+    msg: OwnedMessage,
+    mut decoder: AvroDecoder,
+) -> Result<DatasetEvent, String> {
+    match decoder.decode(msg.payload()) {
+        Ok(result) => match result.name {
+            Some(name) => match name.name.as_str() {
+                "DatasetEvent" => match name.namespace {
+                    Some(namespace) => match namespace.as_str() {
+                        "no.fdk.dataset" => match from_value::<DatasetEvent>(&result.value) {
+                            Ok(event) => Ok(event),
+                            Err(e) => Err(format!("Deserialization failed {}", e)),
+                        },
+                        ns => Err(format!("Unexpected namespace {}", ns)),
+                    },
+                    None => Err("No namespace in schema, while expected".to_string()),
+                },
+                name => Err(format!("Unexpected name {}", name)),
+            },
+            None => Err("No name in schema, while expected".to_string()),
+        },
+        Err(e) => Err(format!("error getting dataset-event: {}", e)),
+    }
+}
+
+// Read DatasetEvent message of type DATASET_HARVESTED
+fn handle_dataset_event(msg: OwnedMessage, decoder: AvroDecoder) -> Result<Option<MQAEvent>, String> {
+    info!("Handle DatasetEvent on message {}", msg.offset());
+
+    let dataset_event = parse_dataset_event(msg, decoder);
+
+    match dataset_event {
+        Ok(event) => {
+            match event.event_type {
+                DatasetEventType::DatasetHarvested => {
+                    let dt = Utc.timestamp_millis(event.timestamp);
+                    info!(
+                        "{} - Processing dataset harvested event with timestamp {:?}",
+                        event.fdk_id, dt
+                    );
+                    parse_rdf_graph_and_calculate_metrics(event.fdk_id, event.graph).and_then(|evt| Ok(Some(evt)))
```

Comment:
Slipper ny `Ok` ved bruk av `and`.
```suggestion
                    parse_rdf_graph_and_calculate_metrics(event.fdk_id, event.graph).map(|evt| Some(evt))
```

Reply:
map mener du?

Reply:
Hehe, ja :sweat_smile: